obspy/clients/fdsn/mass_downloader/download_helpers.py

```diff
@@ -349,19 +349,29 @@ def sanitize_downloads(self, logger):
         of the download helpers does not allow for a StationXML file with no
         data.
         """
+        from obspy.io.mseed.util import get_start_and_end_time
         # All or nothing for each channel.
         for id in self.miss_station_information.keys():
-            logger.warning("No station information could be downloaded for "
-                           "%s.%s.%s.%s. All downloaded MiniSEED files "
+            logger.warning("Station information could not be downloaded for "
+                           "%s.%s.%s.%s. MiniSEED files outside of the "
+                           "station information period "
                            "will be deleted!" % (
                                self.network, self.station, id[0], id[1]))
             channel = [_i for _i in self.channels if
                        (_i.location, _i.channel) == id][0]
             for time_interval in channel.intervals:
+                # Check that the time_interval.start and end are correct!
+                time_interval.start, time_interval.end = \
+                    get_start_and_end_time(time_interval.filename)
                 # Only delete downloaded things!
                 if time_interval.status == STATUS.DOWNLOADED:
-                    utils.safe_delete(time_interval.filename)
-                    time_interval.status = STATUS.DOWNLOAD_REJECTED
+                    # Only delete if the station data are actually missing
+                    # for this time
+                    miss_start, miss_end = self.miss_station_information[id]
```

Comment:
I just had to reread a lot of the code to understand this. So we might have to expand `miss_station_information` a bit to fully get this to work.

`miss_station_information` and its pendant `have_station_information` are per-channel time intervals. It has no concept of partially downloaded channel information - data for each channel is either considered to be completely downloaded or not - I agree that this should be fixed somehow. These things happen in the `download_stationxml` method.

The current solution in this PR will (I think) delete time intervals that are completely inside the complete requested time interval and ignore those at the end as soon as some piece of station information for that particular channel has not been downloaded.

What we should do:

Give the station objects an attribute called `available_station_intervals` (set after the StationXML files have been downloaded) and use that to decide whether or not to discard already downloaded MiniSEED files.

Is my interpretation correct or am I missing something?